imap/jmap_mail.c

```diff
@@ -3858,24 +3858,29 @@ static void emailquery_cache_slice(struct emailquery *q,
                                    json_t **errp)
 {
     if (!qc->have_total) {
-        size_t n = SIZE_MAX;
+        /* conditions for doing a partial load:
+         * 1) we don't need the total
+         * 2) we don't need to find an anchor
+         * 3) the offset is from the start, not from the end
+         * 4) we have a limit, so we're not fetching everything
+         * 5) we don't need all in thread
+         * if all these conditions are met, then we can fill just to the
+         * end of the requested window */
         if (!q->super.calculate_total && !q->super.anchor &&
             q->super.position >= 0 && q->super.limit && !q->findallinthread) {
-            /* Lazily load at most n matches into cache */
-            size_t have_len;
-            if (q->collapse_threads) {
-                have_len = qc->collapsed_len;
-            }
-            else {
-                have_len = qc->uncollapsed_len;
-            }
-            size_t lastpos = q->super.position + q->super.limit;
-            if (lastpos > have_len) {
-                n = lastpos - have_len;
+            size_t have_len = q->collapse_threads ?
+                              qc->collapsed_len :
+                              qc->uncollapsed_len;
+            size_t need = q->super.position + q->super.limit;
+            // is the window already within the loaded section?
+            if (have_len < need) {
+                // fill just to the end of the window
+                emailquery_cache_loadn(q, qc, need);
```

Comment:
The `loadn` function is a misnomer I introduced, and now I can now how it is misleading. A better name might be `ensuren` or even `ensure`, like we do in the buf API.

Reply:
I'll rename it!